src/tests/sqlsmith/develop.md

```diff
@@ -1,35 +1,119 @@
-3. Cleanup phase - Cleanup created relations.
+For instance adding `ORDER BY`, or `Emit On Window Close`.
 
-### Setup phase
+There isn't a clear-cut way for these.
 
-1. `CREATE TABLE` - Created from SQL source files `sqlsmith/tests/testdata`).
-2. `CREATE MATERIALIZED VIEW` - Generate materialized views on seeded tables, previously generated materialized views.
+Typically, to add them you would do the following:
+1. Identify which part of the AST these are from.
+2. Look where this part of AST it is generated.
+   The section on [[Query Generation]] should help.
+   The generation process roughly mirrors the structure of the AST.
+3. Add it to that part of the AST.
 
-### Query phase
+If it affects generation of other parts of the query, you may wish to re-order the generation process.
+For instance, the `GROUP BY` clause affects which columns we can select,
+so we generate it before generating the `SELECT` items.
+
+> TODO: Here provide an example / tutorial of adding `Emit On Window Close`.
+
+## Adding new expressions
+
+To check if your expression is included just generate ft.txt (see README.md), and make sure your newly added function is there.
+
+> TODO: Generate a not-ft.txt which includes functions not tested by sqlsmith.
+
+The following sections are on adding various expression types.
+
+### Functions
+
+Mostly for new functions it should have been generated by default, since we use frontend's type signature map to generate functions.
+
+For functions of variable arity or functions with three or more arguments, they may not be included.
+
+In that case, you may reference the implementation of `concat` for functions of variable arity.
+For functions with fixed arity but more than three args, they do not use vectorized execution,
+and hence they do not have custom implementations like `Expr::Binop` or `Expr::Unop`. Instead, they get translated to a `Function`.
+Their implementation can be referenced via `make_simple_func`.
+
+### Aggregates
+
+New Aggregates should already be included by default.
+
+### Casts
+
+New Casts should already be included by default, or blacklisted explicitly.
+
+## Test Runner overview
+
+Query execution and generation happen in step. Here's an overview of it.
+1. Create Base tables, based on the supplied `test_data` argument to sqlsmith.
+2. Generate `INSERT`s and Populate base tables.
+3. Generate and Create base materialized views, they will be used to test mv-on-mv.
+   These materialized views can either be on Base tables,
+   or on a previously created materialize view from this step.
+4. Generate `UPDATE / DELETE` statements and Update base tables with them.
+   If no PK we will just do `DELETE` for some rows and `INSERT` back statements.
+5. Generate and run batch queries e.g. `SELECT * FROM t`, `WITH w AS ... SELECT * FROM w`.
+6. Generate and run stream queries. 
+   These are immediately removed after they are successfully created.
```

Comment:
Sometimes we may want to test insert/update/delete after MV being created, can we achieve that?

Reply:
These are tested in step 3 and 4. Once base tables are updated, these updates should be propagated to mvs created in step 3.

If we wish to interleave updates/deletes/inserts between each stream query in step 6,
1. The time taken to run the e2e test could greatly increase.
2. It takes extra effort to extract the failing set of queries.

IMO because it is tested in step 3 and 4 we can leave it out for now.
We can always scale up the testing by:
1. Increasing number of base mvs created.
2. Just running more sets of sqlsmith with different test seeds.